src/BenchmarksApps/Kestrel/PlatformBenchmarks/BenchmarkApplication.HttpConnection.cs

```diff
@@ -41,41 +41,50 @@ private async Task ProcessRequestsAsync()
         {
             while (true)
             {
-                var task = Reader.ReadAsync();
-
-                if (!task.IsCompleted)
+                var readResult = await Reader.ReadAsync();
+                
+                if (!HandleRequest(readResult))
                 {
-                    // No more data in the input
-                    await OnReadCompletedAsync();
+                    return;
                 }
 
-                var result = await task;
-                var buffer = result.Buffer;
-                while (true)
+                await Writer.FlushAsync();
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool HandleRequest(in ReadResult result)
+        {
+            var buffer = result.Buffer;
+            var writer = GetWriter(Writer);
```

Comment:
ParseHttpRequest can (at least theoretically) return true without switching `_state` to `State.Body`. If you move this inside `if (_state == State.Body)` it might save on some unneeded initialization.

Also, maybe I'm missing something, but wasn't the BufferWriter initialized once per request/response before this change?

Reply:
> wasn't the BufferWriter initialized once per request/response before this change

It was for a non-pipelined scenario (JSON). For pipelined scenario (Plaintext only) it was executed 16 times (at least to my understanding)